query/src/client/QueryMetadataEditor/QueryMetadataEditor.tsx

```diff
@@ -0,0 +1,374 @@
+            showViewDataConfirmationModal: false
+        };
+    }
+
+    componentDidMount() {
+        const { schemaName, queryName, messages } = this.state;
+
+        if (schemaName && queryName) {
+            fetchQueryMetadata(schemaName, queryName)
+                .then((domain) => this.setState(() => ({domain})))
+                .catch((error) => {
+                    this.setState(() => ({
+                        messages: messages.set(0, {message: error.exception, messageType: 'danger'})
+                    }));
+                });
+        }
+        window.addEventListener("beforeunload", this.handleWindowBeforeUnload);
+    }
+
+    handleWindowBeforeUnload = (event) => {
+        if (this.state.dirty) {
+            event.returnValue = 'Changes you made may not be saved.';
+        }
+    };
+
+    onChangeHandler = (newDomain, dirty) => {
+        this.setState((state) => ({
+            domain: newDomain,
+            dirty: state.dirty || dirty, // if the state is already dirty, leave it as such
+            showSave: true
+        }));
+    };
+
+    showMessage = (message: string, messageType: string, index: number, additionalState?: Partial<IAppState>) => {
+        const { messages } = this.state;
+
+        this.setState(Object.assign({}, additionalState, {
+            messages : messages.set(index, {message: message, messageType: messageType})
+        }));
+    };
+
+    dismissChangeConfirmation = () => {
+        const {showViewDataConfirmationModal, showEditSourceConfirmationModal} = this.state;
+
+        this.setState(() => ({
+            showResetConfirmationModal: false,
+            showViewDataConfirmationModal: false,
+            showEditSourceConfirmationModal: false,
+            dirty: false
+        }));
+
+        if (showViewDataConfirmationModal) {
+            this.navigateToViewData();
+        }
+
+        if (showEditSourceConfirmationModal) {
+            this.navigateToEditSource();
+        }
+    };
+
+    dismissAlert = (index: any) => {
+        this.setState(() => ({
+            messages: this.state.messages.setIn([index], [{message: undefined, messageType: undefined}])
+        }));
+    };
+
+    aliasFieldBtnHandler = () => {
+        this.setState(() => ({showAlias: true}));
+    };
+
+    onHideAliasField = (): any => {
+        this.setState(() => ({showAlias: false}));
+    };
+
+    onAddAliasField = (domainField: DomainField): any => {
+        const { domain } = this.state;
+        const newFields = domain.fields.push(domainField);
+        const newDomain = domain.set('fields', newFields) as DomainDesign;
+
+        this.setState(() => ({
+            showAlias: false,
+            domain: newDomain,
+            showSave: true,
+            dirty: true
+        }));
+    };
+
+    onConfirmEditSource = () => {
+        const { dirty } = this.state;
+
+        if (dirty) {
+            this.onSaveBtnHandler();
+        }
+
+        this.navigateToEditSource();
```

Comment:
Since the this.onSaveBtnHandler() involves an Ajax call, won't this line execute too early (i.e. before the save has happened)? Probably should pass through the callback function to the onSaveBtnHandler. Something like this maybe:
```suggestion
        if (dirty) {
            this.onSaveBtnHandler(this.navigateToEditSource);
        }
        else {
             this.navigateToEditSource();
        }
```